### DISCUSSION TITLE
Dampening

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#120] [Ideas]
I think a beneficial improvement may be interpolated dampening values for inter-hour forecast values.

<img width="268" alt="image" src="https://github.com/user-attachments/assets/9295ad3f-1ba0-4a93-a8ef-14bf47472f19">

As we know, these are currently configured by the hour, which makes total sense. All generally lines up nicely on the Energy dashboard for a clear day, yet falls somewhat short when one utilises the half-hourly forecast attribute values for things like the chart above (thank you, @gcoan).

<img width="250" alt="image" src="https://github.com/user-attachments/assets/54620324-9029-4072-8c37-2b7a0baa7776">

The forecast for each day has the same hourly dampening value applied to both 30-minute intervals, and this strikes me as room for improvement.

Thoughts?

**Comment by gcoan** [comment]
There are pro's and con's of adding interpolation to the dampening output.

Dampening to take account of say shading from next door's house, that shades the panels by 50% up to 9am then 0% from 10am,  an interpolated 25% shading at 9:30am for the half hourly forecast might be correct in that the panels are partially shaded, or it might be wrong that it really is a vertical time line of when the shading stops.

Would depend in part on what type of shading it is, a building giving a much more knife edge to the shading whereas a tree would be softer transition.

(Ignoring of course that the dampening factors should change through the year as the sun elevation changes through the seasons)

Might be the case of deckchairs on the titanic to fiddle with this

**Reply by Nilogax** [reply]
I'm with @gcoan on this. At some point you have to say "OK it's a bit wrong but it's just a forecast" and I'd be interested in a use case where interpolation of dampening is a material improvement over the current method. With my automated dampening updates, most days my generation tracks around +/-5% of my dampened forecast, then some days like yesterday it's way way cloudier and I get around 75%, some days the clouds disappear and I end up at 120%.

Might be more beneficial to the user base to incorporate something like my automated dampening from within the integration? Predbat does something similar with battery charge and discharge curves, either setting them from recent history or allowing the user to override.

**Reply by autoSteve** [reply]
Righto. Fair enough. Interpolation for dampening is silly.

All fair enough, @gcoan. What would not be silly is changing dampening to be per-half-hour to align with the forecasts, but that might get stupid, especially given automated dampening that has already been built.

Your automation I am interested in, @Nilogax. I've not seen it. (Thinks: Switch it on as a by-default-off option?) That is a case where per-half-hour dampening may not be silly...

**Reply by Nilogax** [reply]
It's not exactly elegant at the moment and I'm sure there are better ways of achieving this from within an integration but the mechanics of it are

Hourly automation to record generation to a "day 0" CSV file as hour, generation so far today (runs 04:00 to 23:00 to cover my best possible generation day) 

Midnight automation to call a python script that removes "Day N"  CSV and shuffles the rest along so day 0 becomes day 1 etc 

01:05 automation to reset dampening to all 1s to generate unmodified forecast 

01:08 automation to call python script that calls solcast.query forecast data  to get (now undamped) forecast for last N days. Import each of the N generation CSV files to determine hourly incremental generation, then for each hour divide sum of N days generation by sum of N days forecast to get a dampening factor. (set max to 1, catch divide by zero, fudge around GMT/BST transition and half hearted attempt to trap some cases of missing data). Write dampening factors back to solcast. 

(why automate at 01:xx? In ye olde days before BJReplay calls to set dampening were burning API calls, so I set it to 01:xx so it was always after midnight UTC and I had fresh API quota) 

I currently use 5 days of history which seems about right. Also looked at getting a generation/forecast ratio for each hour and then averaging those figures to get a dampening factor but when forecast per hour is low that gives some wild results so using the overall total is more stable but then if you have a high/low outlier day in the history that would affect the current calculation method. Also my inverter only increments generation in 0.1kwh steps so the numerator isn't particularly reliable at low generation times. Going to a 30 min dampening would exacerbate that issue as the numbers would be even smaller, but agree it kinda makes sense to align dampening intervals with forecast intervals. 

My current method doesn't have much in the way of resilience checking, and obviously I know that I need to turn it off if for example a solcast upgrade nuked my history 😊

To go to 30 min dampening without breaking any existing automations perhaps just check the number of parameters passed to the set dampening service - 24 => hourly dampening so set xx:00 and xx:30 to the same value, 48 => half hourly dampening 

A "nice to have" feature would be some way of surfacing the current dampening settings so they can be displayed or charted. 

As for the user experience I'd suggest 2 modes of operation for dampening - manual and automatic. Manual would calculate proposed figures (output to the log?) but not change anything, automatic would calculate and update. Keep Manual as the default as I'd expect if people aren't using dampening (eg @gcoan?) they have other ways of aligning forecasts vs their achieved generation so adding dampening could break that.

**Reply by gcoan** [reply]
That's quite a neat solution, I like the dynamic adaptation of dampening.

As a build on it, rather than outputting to the log file, output the damping factors to a sensor so then they can be charted etc in home assistant

**Reply by Nilogax** [reply]
Thanks. In terms of output,maybe two sensors, one for current dampening settings in use and one for latest calculated settings, that way in manual mode you could chart current vs calculated and decide whether to make a change. Likewise possibly have "days history" as an input number (default 5?) - longer periods should give more stable calcs but would tend to be more pessimistic (over damping) as shading reduced and optimistic as shading increased..

**Reply by autoSteve** [reply]
I have thought long and hard about this over the past few days.

I see issue with using the generation history. Well, I see other issues as well.

For me, my nasty power distributor (not retailer, who are also nasty) requires that a 5kW export limit be in place for the site. This has the effect that of an afternoon once my batteries fill, the PV export is curtailed if it exceeds the limit. Because 12kW AC generation it _will_ be curtailed _a lot_ in Summer. This could be interpreted as shading.

This led me down a rabbit hole of examining the history of the generation sensor vs. site export sensors to try and determine whether curtailing was happening for a given period. My thought was to exclude these curtailed readings to not impact dampening. Turns out that the sensor does not hit a red-line of 5kW. Sometimes 5.4. Sometime 5.1. Sometimes... Even averaged for half-hour periods the number jumps.

This all led me down a further rabbit hole of, "How does one get sensor history data for a sensor that my integration does not own?" Which led to the creation of a user API token, which got messy from an integration standpoint. It would need to be added in integration configuration. Annoyingly if this integration were an HA add-on instead then I could make supervisor proxied calls to HA without needing an API token (assuming a supervisor was part of the install set-up, which sometimes it is not). An integration cannot just blindly make calls to get other integration sensor values and history.

Then I deleted the dev branch. On purpose.

**Reply by Nilogax** [reply]
Thanks for taking a look. Seems like a good example of why it's so much easier to build a home grown solution that fits your (my?) own needs than to build a solution that will support all the wierd and wonderful configurations out in the wild.

Playing devil's advocate, does this just mean that your setup isn't suitable for automated damping? So if you had calculated damping values available you could determine which ones are valid (shading etc) and which are artefacts of network rules.

**Reply by autoSteve** [reply]
That was the plan. I have failed for now.

I would love to get automated dampening running factoring network limit, esp. in Winter when I do get some shading at the tail end of a day because of a bloody big tree a couple houses down. Spring/Summer/Autumn is epic. Ugh. Back to HA automations for now, bad as they are for me.

I will stow this in the back of my mind to work on an automated integration solution some day.

**Reply by autoSteve** [reply]
BTW, ChatGPT did a super half-arsed job just now, if I'm honest, totally avoiding Python.

`built an automated home assistant solar forecast dampening calculation for BJRaplay Solcast integration that takes into account transmission limit`

The result soooo won't work. And I'm so not surprised. Any wonder SkyNet got it so wrong in Terminator with everyone dying.

The "build it in Python" version was even more hystrical, even though our overlord was confident in its response.

I still have not used ChatGPT for anything at all. And doubt I ever will.

**Reply by autoSteve** [reply]
One last one, sorry. This nearly broke its head, despite prior questioning.

`Build an automated home assistant solar forecast 24-hour dampening calculation in Python, adding it to the Ozzie Solcast integration, taking into account feed-in limit, but also factoring battery state of charge. The dampening needs to take into account the prior five days of actual production.`

It literally stalled for ten seconds before anything came back. Then there was spewed unsurprisingly gibberish-esq Python.

This is a fun game.

But leave it to the grown-ups, ChatGPT, yeah?

**Comment by froxic** [comment]
I have another idea. What about to adjust dampening curve based on historic PV generation from PV HA sensor? Algo then adjust dampening proactively based on e.g. last 7 days generation with correlation to previous SolCast forecast. This could be nice also during whole year since PV generation could differ during autumn / winter period for instance and on my installation i need to readjust dampening based on hours since there is sunset 6 hours sooner.

**Reply by BJReplay** [reply]
> What about to adjust dampening curve based on historic PV generation from PV HA sensor?

That's the idea: you put together an automation that does just that.  You compare your actual output to the last say seven days forecasts and move the dampening up or down.

Bear in mind that Solcast is already forecasting your insolation based on:
- Latitude
- Longitude
- Installation Date (so it will work out panel degradation based on time) to calculate a loss factor, or
- Efficiency Factor
- Azimuth
- Elevation
- Insolation based on time of year and cloud forecasts.

So you shouldn't need to worry about the sun setting six hours earlier, unless you have shading from buildings or trees - but if you do, then use an automation that you write to learn from your last seven days actuals to move your dampening.

**Reply by gcoan** [reply]
I have an automation that adjusts dampening factors based on sunset/sunrise times across the year, to reflect shading moving as the sun moves through the seasons.

Was going to share it as a PR to the integration but could share it here if people want to use it as a start point for more clever shading/dampening coding.

**Reply by BJReplay** [reply]
> Was going to share it as a PR to the integration

I believe that this sort of thing is better off here in a discussion, rather than being - added to the code base:
- if added to the code base, implies also requires support
- presumably requires configuration at least
- how do you configure?
  - is it azimuth? 
  - is it time of day? 
  - is it time of year?

In summary, I don't see a general, widespread solution that can be easily added and supported.

**Reply by gcoan** [reply]
Here's a link to the dampening code I mentioned above https://github.com/springfall2008/batpred/issues/1219#issuecomment-2177812604

Its not my code, I was just suggestion to share it more widely for reuse by the community

**Reply by Nilogax** [reply]
I'll try to find some time to make my solution presentable and document what steps are needed to get it up and running. It was my first attempt at building anything for HA and I went down some blind alleys along the way so need to sort the wheat from the chaff.

**Reply by Nilogax** [reply]
OK, found some time so here's the first step which is the Sunshine Recorder to record actual generation history.  This needs to run for a few days before the damping calcs can start.  What should happen is you end up with a folder /adaptivedamping populated with a set of csv files called pv_day_X.csv where day_0 is the incomplete current day and day 1 is yesterday, day 2 is the day before etc up to the max history you've defined.  These files are then shuffled along each day, with the oldest file discarded.  You'll also need to discard the partial day record that gets created when you set this up, or wait for it to be discarded.  

CAVEATS - this can almost certainly be improved (suggestions welcome), I was learning everything as I went along.  Error checking is minimal, so it will most likely fall over if the wind changes.  More specifically I used utcnow() to log the time after it struggled when we switched to from GMT to BST, but I'm not sure what'd happen if your timezone is UTC+10 or whatever. 

Anyhow... first you'll need the File integration 

https://www.home-assistant.io/integrations/file/

Then create a subfolder /adaptivedamping in /config (where configuration.yaml lives)

Edit configuration.yaml to include

```jinja
homeassistant:
  allowlist_external_dirs:
    - /config/adaptivedamping
```

Reload the YAML configuration then in the File integration configure a new File notification service with path 

```
/config/adaptivedamping/pv_day_0.csv
```

do not select 'Add Timestamp'

Once the File entity has been created, modify the entity name to AdaptiveDamping and change the entity id to

```
notify.adaptivedamping
```

Then create the automation to record PV generation (updated after feedback below)

```yaml
alias: AdaptiveDamping - PV Recorder
description: PV Recorder
trigger:
  - platform: time_pattern
    hours: /1
condition:
  - condition: time
    after: "00:05:00"
action:
  - target:
      entity_id: notify.adaptivedamping
    data:
      message: >-
        {{utcnow().hour}},
        {{states('sensor.YOUR SENSOR THAT GIVES TOTAL PV GENERATION SO FAR TODAY')}}
    action: notify.send_message
mode: single
```

Create a helper that defines how many days history you'll be keeping - I settled on 5 days

```
input_number.adaptivedamping_history
```

Then you'll need to get pyscript from HACS

https://github.com/custom-components/pyscript

Make sure you update your configuration.yaml to allow pyscript to access all libraries (you'll need to reload your HA config again after making this change):

```yaml
pyscript:
  allow_all_imports: true
  hass_is_global: true
```

And create adaptivedamping_movefiles.py in the /config/pyscript folder

```python
@service
def adaptivedamping_movefiles():
    
  import os, sys
    
  history_days = int(float(state.get('input_number.adaptivedamping_history')))
  
  filename_stub = "adaptivedamping/pv_day_"
 
  #Log start action execution
  #log.info("START of MoveFiles")
  
# check if we can delete last file

  filename = filename_stub + str(history_days) +".csv"
  
 
# if last old file exists  delete it
  
  if os.path.isfile (filename):
    os.remove(filename)

# now loop shuffling day n-1 to day n

  for daynum in range(history_days,0, -1):
    if os.path.isfile(filename_stub + str(daynum-1) + ".csv"):
      os.rename (filename_stub + str(daynum-1) + ".csv", filename_stub + str(daynum) + ".csv")
    
  filename = filename_stub +"0.csv"    

  if os.path.isfile (filename):
    os.remove(filename)
    
  #log.info( "AdaptiveDamping END of MoveFiles")  
```
And finally the automation to trigger this script

```yaml
alias: AdaptiveDamping - Move Files
description: ""
trigger:
  - platform: time
    at: "00:05:00"
condition: []
action:
  - data: {}
    action: pyscript.adaptivedamping_movefiles
mode: single
```

Part two to follow (sometime) - that's the bit that actually calculates the dampening factors

**Reply by gcoan** [reply]
Neat logic, am interested to see part 2!

I was thinking about the trigger code, one potential simplification could be to trigger the automation every hour regardless of the time of day.  If you wanted to only record generation when the sun was up, could add a condition that stops the automation if the sun is set, but as written it would miss out solar generation from the last partial hour before sunset - would need some more clever OR logic inserting 

```
trigger:
 - platform: time_pattern
    hours: /1
condition:
  - condition: state
    entity_id: sun.sun
    state: above_horizon
    enabled: true
```

**Reply by Nilogax** [reply]
Thanks - agree the trigger is clumsy and as written would need modifying if eg. someone was up near the Arctic circle and getting generation over longer hours.  I'm not sure why I don't just record it every hour TBH - will have a think and see if I can remember any good reason!  

I've been revisiting the code for part 2 and stripping out some bits that I thought I might need but never used.  Will give it a day or two in case the new version falls over then share it.

**Reply by Nilogax** [reply]
So here's Part 2 and an update to Part 1, following @gcoan's suggestion above.  All feedback, suggestions and critique welcome.  Edit - updated to work with changes to how dampening is applied in Solcast 4.1.9 - will not give sensible results with earlier versions

First the Part 1 update.  I was thinking about why I bothered with the specific times in the automation and the only things that I could come up with were either an early attempt to avoid issues with GMT/BST or some sort of data minimisation attempt which probably saves less data than it takes to store the list of times in the trigger.  So a simple hourly trigger should work apart from getting a bit dodgy around midnight as the 'PV so far today' figure might still be at its 23:59:59 value when the 00:00 PV recording automation triggers.  So I added a condition to skip 00:00 and all seems OK.    Have updated in the post above.  

The second part compares the recorded PV with the historic forecast and recalculates the damping factors.  you'll need to create a couple of helpers for this:

```
input_boolean.adaptivedamping_write_enable
```

Which controls whether the calculated damping settings are written back to Solcast.  Leaving it set to off will still do all the calculations but won't update the Solcast damping.  It will still update the attributes in sensor.adaptivedamping (see below).

The final helper is 

```
input_number.adaptivedamping_delta_scale
```

which should be defined with a range from 0-1.  This controls how much of the difference between the old value and new value is applied to each damping setting - e.g if the old value was 0.6 and the new was 0.8 then setting the delta_scale to 0.25 would mean the new setting was 0.65 i.e 0.6 + 0.25 * (0.8-0.6), setting it to 0.5 would mean the new setting was 0.7 i.e. 0.6 + 0.5*(0.8-0.6).  This is a way of making the damping settings less volatile though you can probably achieve something similar by using a longer history.  I've settled on a 5 day history and a 0.33 delta scale.

Finally the main Python script creates a sensor

```
sensor.adaptivedamping
```

The state of this sensor contains the status reported by the update process, and it has attributes that I use for charting - 'data' which lists the current damping settings and 'delta' which lists the changes.  This sensor won't survive a restart.  Do not attempt to create this independently as it can cause issues (see feedback from CTRDevelopments below)

There is then an automation to complete the recalculation.  Don't enable this until you've got a few days of PV history files from part one above.  

Why trigger at 01:08?  There was a time when set_dampening was burning Solcast API calls so I moved this to always run after midnight GMT, but that's no longer the case so it could be moved earlier (but after the move files automation has run).  The damping update process just needs to complete before anything else needs to use the Solcast forecast data so in my case that's Predbat getting the plan sorted before the 02:00 Octopus Flux rate kicks in.   Remember that the new damping factors won't be applied until a forecast is fetched from Solcast.

```yaml
alias: AdaptiveDamping - Update
description: ""
trigger:
  - platform: time
    at: "01:08:00"
condition: []
action:
  - data: {}
    action: pyscript.adaptivedamping_update
mode: single
```

Then we get the meat of the calculation - in the pyscript folder create a file adaptivedamping_update.py:

```python
@service
def adaptivedamping_update():
    
  from datetime import date
  from datetime import timedelta
  from datetime import datetime, timezone
  from array import array
  import os, sys
  import csv

  hours_in_day = 24    
  
  warning = False

  forecast_total = [0] * hours_in_day
  cumulative_generation = [0] * hours_in_day
  incremental_generation = [0] * hours_in_day
  damping_config = [1] * hours_in_day
  last_config = [1] * hours_in_day
  damping_delta = [0] * hours_in_day
  
  state.set('sensor.adaptivedamping', "FAILED")

  history_days = int(float(state.get('input_number.adaptivedamping_history')))
  
  #preserve previous hourly damping settings
  
  #previous_damping = state.get('input_text.adaptivedamping_config')

  damp_response = solcast_solar.get_dampening(return_response = True)
  
  damp_settings = damp_response.get('data')
  
  damp_count = 0
  
  for key in damp_settings:
      previous_damping = key.get('damp_factor')
      damp_count+=1
  
  if damp_count != 1:
      log.warning ("Adaptive Damping retrieved " + str(damp_count)+ " sites, expected 1")
      warning = True

  preserved_values = previous_damping

  for h in range(hours_in_day-1):
      last_config[h] = float(previous_damping[0:previous_damping.index(',')])
      previous_damping = previous_damping[previous_damping.index(',')+1:]
      
  #Iterate through PV history files to extract actual generation

  daynum = 1
  last_pv_day = 0
  
  while daynum <= history_days:
      filename = "adaptivedamping/pv_day_"+str(daynum)+".csv"
      if os.path.isfile(filename):
          csv_file = os.fdopen(os.open ("adaptivedamping/pv_day_"+str(daynum)+".csv",os.O_RDONLY))
          contents = csv.reader(csv_file)
          line_count = 0
          for line in contents:
              if line_count < 2:
                  line_count += 1
              else:
                  hour = int(line[0])
                  cumulative_generation[hour] += float(line[1])
                  line_count += 1
          if line_count != 25:
              log.warning("Adaptive Damping - PV History for day "+str(daynum)+" contains "+str(line_count-2)+" records")
              warning = True          
          csv_file.close()
          last_pv_day = daynum
      else:
          log.warning("AdaptiveDamping - PV History File not found for day "+str(daynum))
          warning = True
      daynum += 1

  # adjust history length if fewer PV day files exist
  
  if last_pv_day != history_days:
      log.warning("Adapting Damping - Found "+str(last_pv_day)+" PV files vs history_days "+str(history_days))
      warning = True
      history_days = last_pv_day
      
  # get required forecast history from Solcast
  
  today = date.today()
  end_time = today.strftime('%Y-%m-%d') +" 00:00:00"
  
  start = today-timedelta(days=history_days)
  start_time = start.strftime('%Y-%m-%d')+" 00:00:00"
  
  response = solcast_solar.query_forecast_data(start_date_time= start_time, end_date_time= end_time, undampened = True, return_response=True)
  
  forecastinfo = response.get('data')
  
  #Iterate through Solcast data summing half hourly figures and shifting times to UTC
  
  solcast_records = 0
  
  for key in forecastinfo:
    period = key.get('period_start')
    period_utc = period - period.utcoffset()
    period_start = period_utc.hour
    forecast_total[period_start]+=key.get('pv_estimate')/2
    solcast_records += 1
    
  # verify we have correct number of solcast records    

  if solcast_records != (history_days*hours_in_day*2):
      log.warning("Adapive Damping - Solcast returned "+str(solcast_records)+ "items, expected "+str(history_days*hours_in_day*2))
      warning = True
      
  damping_text=""
  
  #Delta Scale determines how much of calculated change is implemented
  
  delta_scale = float(state.get('input_number.adaptivedamping_delta_scale')) 

  #Calculate damping factors from PV history and forecast

  now = datetime.now(timezone.utc).astimezone()
  utc_offset = int(int(now.strftime("%z"))/100)
  
  changed_value = False

  for hour in range(hours_in_day-1):
    tz_hour = hour + utc_offset

    #Attempt to deal with larger UTC offsets
    if tz_hour < 0:
        tz_hour += hours_in_day
    elif tz_hour > hours_in_day-1:
        tz_hour -= hours_in_day
        
    if cumulative_generation[hour+1] >= cumulative_generation[hour]:
        # avoid potential for negative generation
        incremental_generation[hour] = cumulative_generation[hour+1]-cumulative_generation[hour]
        
    if (forecast_total[hour]==0):
        damping_config[tz_hour] = 1
    else:
        raw_value = incremental_generation[hour]/forecast_total[hour]
        raw_delta = raw_value - last_config[tz_hour]
        damping_config[tz_hour] = round(max(0,min(last_config[tz_hour] + (delta_scale * raw_delta),1)),3)
        changed_value = True
  
  # copy damping values to string for passing to Solcast

  for hour in range(hours_in_day-1):
    damping_text = damping_text + "," + str(damping_config[hour])
      
  damping_text = damping_text[1:] + ",1"

  # Revert if nothing has changed - most likely solcast didn't return any data

  if not changed_value:
      damping_text = preserved_values
      damping_status = "ERROR"
  else:
      damping_status= "Calculated"
  
  # only update solcast if enabled, otherwise write preserved values back
  
  write_enable = state.get('input_boolean.adaptivedamping_write_enable')

  if (write_enable=="on"):
      damping_status = "Updated"
  else:
      damping_text = preserved_values

  solcast_solar.set_dampening(damp_factor=damping_text)
   
  if warning:
      damping_status += " with warnings - see log"
  
    #Calculate changes made to damping
  
  for hour in range(hours_in_day-1):
      damping_delta[hour] = damping_config[hour] - last_config[hour]
      
  #Populate a sensor to store current status, config and changes for charting

  state.set('sensor.adaptivedamping', damping_status)
  state.setattr('sensor.adaptivedamping.data', damping_config)
  state.setattr('sensor.adaptivedamping.delta', damping_delta)  

  

```

There is undoubtedly some shonky code in this (probably quite a lot) that could be made more elegant, more robust and generally nicer but as mentioned before I was learning Python as I went along so stopped at the first working solution.  

Finally some apexcharts to show the current config and the most recent change

![image](https://github.com/user-attachments/assets/377f994f-f94e-43ef-a801-e752adfbc1ed)


```yaml
- type: custom:apexcharts-card
  experimental:
    color_threshold: true
  header:
    show: true
    title: Damping Factors
    show_states: false
  apex_config:
    legend:
      show: false
    tooltip:
      'y':
        formatter: |
          EVAL:function(y) {return y.toFixed(3);}
      enabled: true
      marker:
        show: false
  graph_span: 23h
  span:
    start: day
  yaxis:
    - min: 0
      max: 1
      decimals: 3
      apex_config:
        tickAmount: 5
  series:
    - entity: sensor.adaptivedamping
      data_generator: >
        let res = []; for (const [key, value] of
        Object.entries(entity.attributes.data)) { res.push([new
        Date().setHours(key,0,0,0), value]); } return res;
      stroke_width: 0
      name: Factor
      type: column
      color_threshold:
        - value: 0.2
          color: darkred
        - value: 0.4
          color: darkorange
        - value: 0.5
          color: gold
        - value: 0.6
          color: forestgreen
        - value: 0.8
          color: royalblue
        - value: 0.9
          color: indigo
        - value: 0.98
          color: darkslategrey
- type: custom:apexcharts-card
  experimental:
    color_threshold: true
  header:
    show: true
    title: Damping Changes
    show_states: false
  apex_config:
    legend:
      show: false
    tooltip:
      'y':
        formatter: |
          EVAL:function(y) {return y.toFixed(3);}
      enabled: true
      marker:
        show: false
    annotations:
      position: back
      yaxis:
        - 'y': 0
          strokeDashArray: 0
          borderColor: white
          borderWith: 2
  graph_span: 23h
  span:
    start: day
  yaxis:
    - decimals: 3
      apex_config:
        forceNiceScale: true
  series:
    - entity: sensor.adaptivedamping
      data_generator: >
        let res = []; for (const [key, value] of
        Object.entries(entity.attributes.delta)) { res.push([new
        Date().setHours(key,0,0,0), value]); } return res;
      stroke_width: 0
      name: Change
      type: column
      color_threshold:
        - value: -1
          color: darkred
        - value: 0
          color: forestgreen
```

**Reply by CTRDevelopments** [reply]
@Nilogax 
> ```
> sensor.adaptivedamping
> ```

Hi, how do i create this? When I run the python script I dont seem to get this sensor created?

cheers

**Reply by Nilogax** [reply]
@CTRDevelopments sorry I'm away at the moment so limited to using my phone to check, but sensor.adaptivedamping should be created on the fly by pyscript when it runs the last 3 lines of the AdaptiveDamping_update script.  Have you set up the solar generation recorder (part one above) and is that working OK? Anything in the logs when you run AdaptiveDamping_update manually?

**Reply by CTRDevelopments** [reply]
Hi, thanks for replying :)
I am getting entries into the file
<img width="524" alt="image" src="https://github.com/user-attachments/assets/cd8a820b-7da6-452f-a4a5-625a14aaef76">

It looks like I am getting an error when the update scripte runs

<img width="806" alt="image" src="https://github.com/user-attachments/assets/74f51941-3a40-4b3f-baae-d637a6afdb82">

Any thoughts?

Cheers

**Reply by BJReplay** [reply]
Turning on debugging for pyscript might help - pyscript will dump out lots of useful information - it helped me find an issue with pyscript (for a completely different task). It dumps out each line.

**Reply by CTRDevelopments** [reply]
> Turning on debugging for pyscript might help - pyscript will dump out lots of useful information - it helped me find an issue with pyscript (for a completely different task). It dumps out each line.

BJReplay I have turned on debug logging for pyscript. Where does it dump out the data to? Cheers

**Reply by BJReplay** [reply]
> BJReplay I have turned on debug logging for pyscript. Where does it dump out the data to? Cheers

It dumps to the same place your home assist logs to (depends on your install type of HA whether it is HAOS or docker or otherwise) - current log is usually called home-assistant.log in the home assistant install directory, and previous log is usually called home-assistant.log.1.  When you turn off debug logging HA prompts you to download the latest debug log as well, so you can try the failing step, and you usually get prompted to save the dump with the diags.

**Reply by CTRDevelopments** [reply]
aha, I think I have fixed it. I had not selected 'allow all imports' in the configuration. It now errors saying it cannot find data for the different days, but that is probably right as I can only see a single data file. I think it should have made a second last night, but I will check again tomorrow. Cheers

**Reply by Nilogax** [reply]
D'oh mea culpa! Poor documentation, sorry.  I'll update it once I'm back home. 

If you've been running the recorder for a few days then I expect the move files script may have been failing for the same reason which is why you only have one file. To start with I'd suggest just running the recorder without updating the damping (or turn off write_enable) until you have at least 3 days history then try updating the damping. Also if you don't have any damping configured at the moment then I'd start with the AdaptiveDamping delta set high (0.75-0.9 range) so the damping factors move toward the calculated values quickly, then drop it down to around 0.33 after a couple of updates so the damping becomes a bit less volatile.  

Please keep me posted with how you get on. 

Edit - pyscript config updated in 'Part 1' post above

**Reply by CTRDevelopments** [reply]
Hi, I have checked this morning and I think i still need a bit of help

I assumed looking at the code that each day it would create another logfile, but I seem to have a single file but with multiple entries?

<img width="503" alt="image" src="https://github.com/user-attachments/assets/033b9e30-4deb-486f-811d-25cf004e6fdb">

Is this right?

**Reply by Nilogax** [reply]
No it always logs to the same file but the adaptive damping move files script should be triggered by an automation at 00:05 each day and rename the files so day_0 becomes day_1 etc.  Suggest checking the traces of that automation and the logs to see if there are any errors.

**Reply by CTRDevelopments** [reply]
this is strange, the automation trace shows it working, but I didnt get the additional file. I have manually run the automation and it does create the new file etc. I have now got a day_0,1,2,3,4 - but the data no good is as clearly is somewhat made up! I will see if the automation triggers tonight and does its magic

**Reply by CTRDevelopments** [reply]
OK, it seems to have worked this morning which is super. However I got an error when the update script ran

<img width="469" alt="image" src="https://github.com/user-attachments/assets/9a1fe2fe-c4b9-4f92-ad59-e14297dd20ff">

Due to my data issues, I had previously copied the same data to all of the logs. (I have disabled the write function while I get this working)

The plots of the damping factors etc are also blank as if it stopped running the script before it got to that stage

**Reply by CTRDevelopments** [reply]
I think I have fixed it. I had previously tried to define a sensor named 'adaptivedamping'. I removed this and then manually ran the code again. It now seems to complete and the graphs update. I will leave it for a few days to collect sensible data now and then update :) thanks for all the help, I have been wanting to do this for a while

**Reply by CTRDevelopments** [reply]
Does it matter if the log file seems to log 1am at the end?

<img width="478" alt="image" src="https://github.com/user-attachments/assets/661361a9-7842-49c4-a530-573a14f78e2d">

**Reply by Nilogax** [reply]
No the log file records times as UTC then the damping update script adjusts for the current time offset

**Reply by CTRDevelopments** [reply]
Had this running now for a few days and it seems to be working nicely. Many thanks for support and a great bit of code! Will continue to see how it goes over the next few weeks

**Comment by Nilogax** [comment]
In the unlikely event that anyone is brave/daft enough to try to use my MacGyver soluton I've updated it so it will now work with Solcast 4.1.9.

**Comment by Nilogax** [comment]
@gcoan I saw your comment in the 4.2.0 discussion re granular damping.   I'm still using hourly damping factors at the moment because of a couple of things.  

First off, there's the combination of my relatively modest generation combined with GivTCP reporting generation in 0.1Kwh steps.  I think breaking down to 30 min intervals should be a pretty simple tweak but I am concerned it could lead to some spurious results.  Maybe using a longer history would help here or maybe adding something like @autoSteve's spliney goodness to the generation figures, maybe it's all in my head and I should just try it anyway.    

Secondly, as the saying goes "It's only a forecast".  Hourly dampening figures are giving results that I feel are acceptable.  On days where the weather follows the forecast I'm getting +/- 10% of the predicted Solcast generation.  Days with bigger variance usually correlate with days where it's unexpectedly sunny (or cloudy) at my peak generation times.   Obviously with higher generation that %age variance translates into more KWh so it could be more important to try to narrow it down with half hourly dampening figures, but the weather will always have the final say.

Also FWIW I have Solcast's efficiency turned right up so I start with an unfiltered forecast and the dampening calcs do all the work.  I didn't see the point in massaging figures that had already been massaged.

**Reply by gcoan** [reply]
Makes sense about turning off the Solcast efficiency rating so you are starting with raw Solcast figures.

I had my efficiency as 98 or 99%, reflecting what I thought would be the case for 18 month old panels.  The main roof is completely unshaded so I expected to get near perfect alignment of forecast to generation, but I didn't.  Part of it was due to having a large number of panels that I push the grid voltage up too high and two of my three inverters will stop generating and then restart a short time later (seconds not minutes). But this throttling off and on knocks the generation back.

Dampening might be a route to getting a more accurate forecast, but given the throttling off and on is power output related, it might just be a different blunt instrument!

For clarification, does your script just operate at the total PV output level or does it work on a per-site basis?  I definitely get more throttling on my main East facing array (site).

But as solar generation is dropping off this is much less of an issue now than it was in the summer.

**Reply by Nilogax** [reply]
Ah sounds like you need some sort of threshold damping so have one value if the forecast PV for the period is above X and another if its below. That's beyond my simple solution for sure but would obviously be more sensitive with 30 min segments. 

Anyhow yes it's built around a single site as that's what I have, so uses total PV generation.  Extending to multiple sites is left as an exercise for the reader 😁

**Reply by erepeo** [reply]
Are there any plans for this project to have automatic hourly damping calculation implemented?

**Reply by autoSteve** [reply]
TL>DR: There are no plans, @erepeo.

Attempting to do this in-integration would require getting generation actuals from outside of the integration to measure against expected forecasts, and then building a weekly/seasonal model and adjusting it over time. We do forecasts. Inverters do actuals. Getting inverter data is possible from the integration, but unweildy and impractical.

The focus of this integration for now is on being rock-solid, providing as much transparency as possible, and as much raw forecast detail as needed to support automations that folks can think up. Actions are available to set dampening via automations / other things at both overall and per rooftop site level. Actions are available to get forecasts, again both overall and per-site. Hourly dampening is a thing. Half-hourly is a thing.

An horizon obstruction model is another thought. If one could define that a local obstruction existed of a certain height, and at a certain distance, width, density and azimuth from "here" it could be possible to determine its impact based on a calculated position of the sun at any given hour of the year, right? Nope. PV generation doesn't work that way. Reflective generation would need to be approximated base on cloud cover, which we don't have, so we would get the answer horribly wrong on mostly cloudy days when diffuse light is king and obstructive trees mean nothing. But Solcast have that answered, just not for us.

Solcast do model irradiance far better than this integration ever could, and for paying customers they do have horizon modelling available. We just consume the more-than-good-enough data produced by their data scientist's typical models. For free. Which is more than good enough for hobbyist applications.

Bottom line, there are no plans from me.

**Reply by erepeo** [reply]
It's a pity but I understand. Thank you for your work.
